src/InterpreterEngine.cpp

```diff
@@ -0,0 +1,1208 @@
+    }
+    auto* program = tUnit.getProgram()->getMain();
+    auto entry = generator.generateTree(*program);
+    InterpreterContext ctxt;
+    if (!profileEnabled) {
+        InterpreterContext ctxt;
+        execute(entry.get(), ctxt);
+    } else {
+        ProfileEventSingleton::instance().setOutputFile(Global::config().get("profile"));
+        // Prepare the frequency table for threaded use
+        visitDepthFirst(*program, [&](const RamTupleOperation& node) {
+            if (!node.getProfileText().empty()) {
+                frequencies.emplace(node.getProfileText(), std::map<size_t, size_t>());
+            }
+        });
+        // Enable profiling for execution of main
+        ProfileEventSingleton::instance().startTimer();
+        ProfileEventSingleton::instance().makeTimeEvent("@time;starttime");
+        // Store configuration
+        for (const auto& cur : Global::config().data()) {
+            ProfileEventSingleton::instance().makeConfigRecord(cur.first, cur.second);
+        }
+        // Store count of relations
+        size_t relationCount = 0;
+        visitDepthFirst(*program, [&](const RamCreate& create) {
+            if (create.getRelation().getName()[0] != '@') {
+                ++relationCount;
+                reads[create.getRelation().getName()] = 0;
+            }
+        });
+        ProfileEventSingleton::instance().makeConfigRecord("relationCount", std::to_string(relationCount));
+
+        // Store count of rules
+        size_t ruleCount = 0;
+        visitDepthFirst(*program, [&](const RamQuery& rule) { ++ruleCount; });
+        ProfileEventSingleton::instance().makeConfigRecord("ruleCount", std::to_string(ruleCount));
+
+        InterpreterContext ctxt;
+        execute(entry.get(), ctxt);
+        ProfileEventSingleton::instance().stopTimer();
+        for (auto const& cur : frequencies) {
+            for (auto const& iter : cur.second) {
+                ProfileEventSingleton::instance().makeQuantityEvent(cur.first, iter.second, iter.first);
+            }
+        }
+        for (auto const& cur : reads) {
+            ProfileEventSingleton::instance().makeQuantityEvent(
+                    "@relation-reads;" + cur.first, cur.second, 0);
+        }
+    }
+    SignalHandler::instance()->reset();
+}
+void InterpreterEngine::executeSubroutine(const std::string& name, const std::vector<RamDomain>& args,
+        std::vector<RamDomain>& ret, std::vector<bool>& err) {
+    InterpreterContext ctxt;
+    ctxt.setReturnValues(ret);
+    ctxt.setReturnErrors(err);
+    ctxt.setArguments(args);
+
+    auto entry = generator.generateTree(tUnit.getProgram()->getSubroutine(name));
+    execute(entry.get(), ctxt);
+}
+
+RamDomain InterpreterEngine::execute(const InterpreterNode* node, InterpreterContext& ctxt) {
+#define DEBUG(Kind) std::cout << "Running Node: " << #Kind << "\n";
+
+#define CASE(Kind)     \
+    case (I_##Kind): { \
+        const auto* cur = static_cast<const Ram##Kind*>(node->getShadow());
```

Comment:
I can do macro overload
```C
#define CASE_CAST(Kind) \
    case (I_##Kind): {  \
        const auto* cur = static_cast<const Ram##Kind*>(node->getShadow());

#define CASE_NO_CAST(Kind, X) case (I_##Kind): {
#define CASE_HANDLER(_1, _2, FUNC, ...) FUNC
/*
 * Usage: For Node type that needs a casting, use CASE(NodeType).
 * For Node type that does not need a casting, provide one extra argument to CASE (e.g. CASE(NodeType, NO_CAST))
 */
#define CASE(...) CASE_HANDLER(__VA_ARGS__, CASE_NO_CAST, CASE_CAST)(__VA_ARGS__)
```
Or just use two seperate marco, `CASE_CAST` and `CASE_NO_CAST`.

Or maybe even get rid of the macro, since using it makes the indention tricky with Clang-Format.

I'm really not sure which one is better.